components/key-exchange/src/exchange.rs

```diff
@@ -63,405 +76,429 @@ pub struct KeyExchangeCore<PS, PR, E> {
+    // to ensure the shares are equal.
+    async fn compute_pms_with(
         &mut self,
-        pms_share1: P256,
-        pms_share2: P256,
+        share_0: P256,
+        share_1: P256,
     ) -> Result<Pms, KeyExchangeError> {
-        let state = std::mem::replace(&mut self.state, State::Error);
-
-        let State::KeyExchange {
-            share_a,
-            share_b,
-            share_c,
-            share_d,
+        let State::Preprocessed {
+            share_a0,
+            share_b0,
+            share_a1,
+            share_b1,
+            pms_0,
             pms_1,
-            pms_2,
             eq,
-        } = state
+        } = self.state.take()
         else {
-            todo!()
+            return Err(KeyExchangeError::state("not in preprocessed state"));
         };
 
-        let pms_share1: [u8; 32] = pms_share1
+        let share_0_bytes: [u8; 32] = share_0
             .to_be_bytes()
             .try_into()
             .expect("pms share is 32 bytes");
-        let pms_share2: [u8; 32] = pms_share2
+        let share_1_bytes: [u8; 32] = share_1
             .to_be_bytes()
             .try_into()
             .expect("pms share is 32 bytes");
 
         match self.config.role() {
             Role::Leader => {
-                self.executor.assign(&share_a, pms_share1)?;
-                self.executor.assign(&share_c, pms_share2)?;
+                self.executor.assign(&share_a0, share_0_bytes)?;
+                self.executor.assign(&share_a1, share_1_bytes)?;
             }
             Role::Follower => {
-                self.executor.assign(&share_b, pms_share1)?;
-                self.executor.assign(&share_d, pms_share2)?;
+                self.executor.assign(&share_b0, share_0_bytes)?;
+                self.executor.assign(&share_b1, share_1_bytes)?;
             }
         }
 
         self.executor
             .execute(
                 build_pms_circuit(),
-                &[share_a, share_b, share_c, share_d],
-                &[pms_1.clone(), pms_2, eq.clone()],
+                &[share_a0, share_b0, share_a1, share_b1],
+                &[pms_0.clone(), pms_1, eq.clone()],
             )
             .await?;
 
-        #[cfg(feature = "tracing")]
-        tracing::event!(tracing::Level::DEBUG, "Successfully executed PMS circuit!");
-
-        let mut outputs = self.executor.decode(&[eq]).await?;
-
-        let eq: [u8; 32] = outputs.remove(0).try_into().expect("eq is 32 bytes");
+        let eq: [u8; 32] = self
+            .executor
+            .decode(&[eq])
+            .await?
+            .pop()
+            .expect("output 0 is eq")
+            .try_into()
+            .expect("eq is 32 bytes");
 
         // Eq should be all zeros if pms_1 == pms_2
         if eq != [0u8; 32] {
-            return Err(KeyExchangeError::CheckFailed);
+            return Err(KeyExchangeError::new(
+                ErrorKind::ShareConversion,
```

Comment:
Ahh I see, I would have considered it its own error kind like `ErrorKind::UnequalPms` to stress that this might be a cheating attempt. Yeah but I guess it works both ways.